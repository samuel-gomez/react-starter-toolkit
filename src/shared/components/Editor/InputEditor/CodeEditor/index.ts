export { default, useCodeEditor } from './CodeEditor';
export type { TReturnUseCodeEditor } from './CodeEditor';
