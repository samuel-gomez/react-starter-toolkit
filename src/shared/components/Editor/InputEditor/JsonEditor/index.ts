export { default, useJsonEditor } from './JsonEditor';
export type { TReturnUseJsonEditor } from './JsonEditor';
