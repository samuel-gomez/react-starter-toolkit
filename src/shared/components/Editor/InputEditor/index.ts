export { default } from './InputEditor';
export type { TInputEditor } from './InputEditor';
