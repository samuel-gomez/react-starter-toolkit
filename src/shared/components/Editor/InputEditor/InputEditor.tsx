import { Text, CheckboxItem, CheckboxModes, Select } from '@axa-fr/react-toolkit-all';
import { DEFAULT_OPTION_LABEL } from 'shared/constants';
import CodeEditor from './CodeEditor';
import JsonEditor from './JsonEditor';
import { Tlistelements } from './CodeEditor/Templates';
import { TonChange } from '../Editor';

type TListSelect = {
  value: string;
  options?: Record<string, string>;
  type?: string;
  list?: Tlistelements;
  labelBtnOpenCodeEditor?: string;
};

export type TInputEditor = {
  value: string | boolean | TListSelect;
  name: string;
  onChange: TonChange;
  options?: { label: string; value: string }[];
};

const commonProps = ({ onChange, name }: Omit<TInputEditor, 'value'>) => ({
  onChange,
  name,
  id: name,
});

type TInputEditorObject = Omit<TInputEditor, 'value'> & {
  value: TListSelect;
};
const InputEditorObject = ({ value, ...props }: TInputEditorObject) => (
  <>
    {(() => {
      switch (true) {
        case !!value.type && value.type === 'separator':
          return <h3 className="af-form-editor__separator">{value.value}</h3>;
        case !!value.type && value.type === 'jsx':
          return <CodeEditor {...commonProps(props)} labelBtnOpenCodeEditor={value.labelBtnOpenCodeEditor} list={value?.list} value={value.value} />;
        case !!value.type && value.type === 'json':
          return <JsonEditor {...commonProps(props)} labelBtnOpenCodeEditor={value.labelBtnOpenCodeEditor} value={value.value} />;
        default:
          return (
            <div className="af-form__select">
              <Select
                {...commonProps(props)}
                forceDisplayPlaceholder={true}
                placeholder={DEFAULT_OPTION_LABEL}
                options={value.options}
                value={value.value}
                aria-label={`select-${props.name}`}
              />
            </div>
          );
      }
    })()}
  </>
);

const InputEditor = ({ value, ...props }: TInputEditor) => (
  <>
    {(() => {
      switch (true) {
        case typeof value === 'boolean':
          return (
            <CheckboxItem
              {...commonProps(props)}
              disabled={false}
              isChecked={value}
              value={`${value}`}
              mode={CheckboxModes.toggle}
              className="af-form__checkbox-toggle"
            />
          );

        case typeof value === 'object':
          return <InputEditorObject value={value as TListSelect} {...props} />;

        default:
          return <Text {...commonProps(props)} type="text" value={`${value}`} />;
      }
    })()}
  </>
);

export default InputEditor;
