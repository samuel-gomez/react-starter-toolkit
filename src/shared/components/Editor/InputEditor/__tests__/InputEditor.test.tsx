import { render } from '@testing-library/react';
import InputEditor from '../InputEditor';

describe('InputEditor', () => {
  const onChange = jest.fn();
  it.each`
    value
    ${true}
    ${'value'}
    ${{ value: 'value', options: [] }}
    ${{ value: 'value', type: 'jsx', labelBtnOpenCodeEditor: 'Edit JSX' }}
    ${{ value: 'value', type: 'json', labelBtnOpenCodeEditor: 'Edit Json' }}
    ${{ value: 'value', type: 'separator' }}
  `('Should render when value: $value', ({ value }) => {
    const { asFragment } = render(<InputEditor value={value} name="name" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
