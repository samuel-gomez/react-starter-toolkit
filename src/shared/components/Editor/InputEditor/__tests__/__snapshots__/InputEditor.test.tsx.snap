// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputEditor Should render when value: {"labelBtnOpenCodeEditor": "Edit JSX", "type": "jsx", "value": "value"} 1`] = `
<DocumentFragment>
  <button
    class="btn af-btn af-btn--editor"
    type="submit"
  >
    <span
      class="af-btn__text"
    >
      Edit JSX
    </span>
  </button>
</DocumentFragment>
`;

exports[`InputEditor Should render when value: {"labelBtnOpenCodeEditor": "Edit Json", "type": "json", "value": "value"} 1`] = `
<DocumentFragment>
  <button
    class="btn af-btn af-btn--editor"
    type="submit"
  >
    <span
      class="af-btn__text"
    >
      Edit Json
    </span>
  </button>
</DocumentFragment>
`;

exports[`InputEditor Should render when value: {"options": [Array], "value": "value"} 1`] = `
<DocumentFragment>
  <div
    class="af-form__select"
  >
    <div
      class="af-form__select-container"
    >
      <select
        aria-label="select-name"
        class="af-form__input-select"
        id="name"
        name="name"
      />
      <span
        aria-controls="name"
        class="glyphicon glyphicon-menu-down"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputEditor Should render when value: {"type": "separator", "value": "value"} 1`] = `
<DocumentFragment>
  <h3
    class="af-form-editor__separator"
  >
    value
  </h3>
</DocumentFragment>
`;

exports[`InputEditor Should render when value: true 1`] = `
<DocumentFragment>
  <div
    class="af-form__checkbox-toggle"
  >
    <input
      checked=""
      class="af-form__input-checkbox"
      id="name"
      mode="toggle"
      name="name"
      type="checkbox"
    />
    <label
      class="af-form__label"
      for="name"
    >
      <span
        class="af-form__indicator"
      >
        <i
          class="glyphicon glyphicon-ok"
        />
      </span>
      <span
        class="af-form__description"
      />
    </label>
  </div>
</DocumentFragment>
`;

exports[`InputEditor Should render when value: value 1`] = `
<DocumentFragment>
  <input
    class="af-form__input-text"
    id="name"
    name="name"
    type="text"
    value="value"
  />
</DocumentFragment>
`;
